src/main/scala/streamer/AdressGenerationUnit.scala

```diff
@@ -0,0 +1,195 @@
+  val loop_counters = WireInit(
+    VecInit(Seq.fill(temporalLoopDim)(0.U(temporalLoopBoundWidth.W)))
+  )
+
+  // siganls indicating can generating next address
+  val addr_gen_counter_inc = Wire(Bool())
+
+  // State declaration
+  val sIDLE :: sBUSY :: Nil = Enum(2)
+  val cstate = RegInit(sIDLE)
+  val nstate = WireInit(sIDLE)
+
+  // Changing states
+  cstate := nstate
+
+  chisel3.dontTouch(cstate)
+  switch(cstate) {
+    is(sIDLE) {
+      when(config_valid) {
+        nstate := sBUSY
+      }.otherwise {
+        nstate := sIDLE
+      }
+
+    }
+    is(sBUSY) {
+      when(addr_gen_finish) {
+        nstate := sIDLE
+      }.otherwise {
+        nstate := sBUSY
+      }
+    }
+  }
+
+  // when cofig valid, store the configuration for later address generation
+  config_valid := io.temporalLoopBounds_i.fire && io.temporalStrides_i.fire && io.ptr_i.fire
+
+  when(config_valid) {
+    temporalLoopBounds := io.temporalLoopBounds_i.bits
+    temporalStrides := io.temporalStrides_i.bits
+    ptr := io.ptr_i.bits
+  }
+
+  // generating next address
+  chisel3.dontTouch(addr_gen_counter_inc)
+  addr_gen_counter_inc := io.ptr_o.ready === 1.B && cstate === sBUSY && !addr_gen_finish
+
+  when(addr_gen_counter_inc === 1.B) {
+    addr_gen_counter := addr_gen_counter + 1.U
+  }.elsewhen(cstate === sIDLE) {
+    addr_gen_counter := 0.U
+  }.otherwise {
+    addr_gen_counter := addr_gen_counter
+  }
+
+  addr_gen_finish := addr_gen_counter === temporalLoopBounds.reduce(
+    _ * _
+  ) && cstate === sBUSY
+
+  // function for generating sub-counters for each nested loop to generate the output address
+  def genNestedLoopCounter(
+      valid: Bool,
+      temporalLoopDim: Int,
+      addr_gen_counter: UInt,
+      temporalLoopBounds: Vec[UInt]
+  ): Vec[UInt] = {
+    val loop_counters = WireInit(
+      VecInit(Seq.fill(temporalLoopDim)(0.U(temporalLoopBoundWidth.W)))
+    )
+
+    // spatailly unrolling the sub loop counter compuration process
+    val addr_gen_counter_next_loop = WireInit(
+      VecInit(
+        Seq.fill(temporalLoopDim + 1)(
+          0.U((temporalLoopDim * temporalLoopBoundWidth).W)
+        )
+      )
+    )
+
+    addr_gen_counter_next_loop(0) := addr_gen_counter
+
+    // Calculate sub loop counters
+    when(valid) {
+      for (i <- 0 until temporalLoopDim) {
+        loop_counters(i) := addr_gen_counter_next_loop(i) % temporalLoopBounds(
+          i
+        )
+        addr_gen_counter_next_loop(i + 1) := addr_gen_counter_next_loop(
+          i
+        ) / temporalLoopBounds(i)
+      }
+    }
+
+    loop_counters
+  }
+
+  // generating sub-loop counetrs using the addr_gen_counter
+  loop_counters := genNestedLoopCounter(
+    io.ptr_o.valid,
+    temporalLoopDim,
+    addr_gen_counter,
+    temporalLoopBounds
+  )
+
+  // address generation with affine mapping definition
+  io.ptr_o.bits := (loop_counters
+    .zip(temporalStrides)
+    .map { case (a, b) => a * b })
+    .reduce(_ +& _) +& ptr
+
```

Comment:
Instead of calculating the correct address with a multiplication every time, does it not make more sense to have a register that is incremented with the stride instead of +1?